substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrFeature.java

```diff
@@ -114,11 +121,35 @@ public void beforeAnalysis(Feature.BeforeAnalysisAccess access) {
 
     @Override
     public void beforeCompilation(BeforeCompilationAccess a) {
-        BeforeCompilationAccessImpl access = (BeforeCompilationAccessImpl) a;
-        int typeCount = access.getTypes().size();
+
+        Map<Class<?>, Integer> classToIndex = new HashMap<>();
+
+        // Scan all classes and build sets of packages, modules and class-loaders. Count all items.
+        Collection<? extends SharedType> types = ((FeatureImpl.CompilationAccessImpl) a).getTypes();
+        int mapSize = 1; // Reserve slot 0 for error-catcher.
+        for (SharedType type : types) {
+            DynamicHub hub = type.getHub();
+            classToIndex.put(hub.getHostedJavaClass(), hub.getTypeID());
+            mapSize++;
+        }
+
+        // Create trace-ID map with fixed size.
+        JfrTraceIdMap map = new JfrTraceIdMap(mapSize);
+        ImageSingletons.lookup(JfrRuntimeAccess.class).setTraceIdMap(map);
```

Comment:
Yes, and if that were possible, I could skip the JfrRuntimeAccess altogether, and register the JfrTraceIdMap directly. However, it seems that this is not possible because when I put the ImageSingletons.add(JfrRuntimeAccess.class) in beforeCompilation() I get complaints by the analyzer that it cannot find the JfrRuntimeAccess, but I don't know about the number of classes before that.

Reply:
ImageSingletons need to be registered early (`beforeCompilation` is too late). But it should be possible to register the uninitialized image singleton in `afterRegistration`.